pkg/ensurance/executor/cpu_usage.go

```diff
@@ -0,0 +1,258 @@
+					klog.KObj(pod), v.ContainerName, containerCPUQuotaNew*containerCPUPeriod.Value)
+
+				released = ConstructCpuUsageRelease(ThrottleDownPods[index], containerCPUQuotaNew, v.Value)
+				klog.V(6).Infof("For pod %s, container %s, release %f cpu usage", ThrottleDownPods[index].PodKey.String(), container.Name, released[CpuUsage])
+
+				totalReleasedResource.Add(released)
+			}
+		}
+	}
+	return
+}
+
+func restoreOnePodCpu(ctx *ExecuteContext, index int, ThrottleUpPods ThrottlePods, totalReleasedResource *ReleaseResource) (errPodKeys []string, released ReleaseResource) {
+	pod, err := ctx.PodLister.Pods(ThrottleUpPods[index].PodKey.Namespace).Get(ThrottleUpPods[index].PodKey.Name)
+	if err != nil {
+		errPodKeys = append(errPodKeys, "not found ", ThrottleUpPods[index].PodKey.String())
+		return
+	}
+
+	// Restore for CPU metric
+	for _, v := range ThrottleUpPods[index].ContainerCPUUsages {
+
+		// pause container to skip
+		if v.ContainerName == "" {
+			continue
+		}
+
+		klog.V(6).Infof("ThrottleExecutor restore container %s/%s", klog.KObj(pod), v.ContainerName)
+
+		containerCPUQuota, err := podinfo.GetUsageById(ThrottleUpPods[index].ContainerCPUQuotas, v.ContainerId)
+		if err != nil {
+			errPodKeys = append(errPodKeys, err.Error(), ThrottleUpPods[index].PodKey.String())
+			continue
+		}
+
+		containerCPUPeriod, err := podinfo.GetUsageById(ThrottleUpPods[index].ContainerCPUPeriods, v.ContainerId)
+		if err != nil {
+			errPodKeys = append(errPodKeys, err.Error(), ThrottleUpPods[index].PodKey.String())
+			continue
+		}
+
+		container, err := utils.GetPodContainerByName(pod, v.ContainerName)
+		if err != nil {
+			errPodKeys = append(errPodKeys, err.Error(), ThrottleUpPods[index].PodKey.String())
+			continue
+		}
+
+		var containerCPUQuotaNew float64
+		if utils.AlmostEqual(containerCPUQuota.Value, -1.0) || utils.AlmostEqual(containerCPUQuota.Value, 0.0) {
+			continue
+		} else {
+			containerCPUQuotaNew = containerCPUQuota.Value / containerCPUPeriod.Value * (1.0 + float64(ThrottleUpPods[index].CPUThrottle.StepCPURatio)/MaxRatio)
+		}
+
+		if limitCPU, ok := container.Resources.Limits[v1.ResourceCPU]; ok {
+			if float64(limitCPU.MilliValue())/CpuQuotaCoefficient < containerCPUQuotaNew {
+				containerCPUQuotaNew = float64(limitCPU.MilliValue()) / CpuQuotaCoefficient
+			}
+		} else {
+			usage, hasExtRes := utils.GetExtCpuRes(container)
+			if hasExtRes {
+				containerCPUQuotaNew = float64(usage.MilliValue()) / CpuQuotaCoefficient
+			}
+			if !hasExtRes && containerCPUQuotaNew > MaxUpQuota*containerCPUPeriod.Value/CpuQuotaCoefficient {
+				containerCPUQuotaNew = -1
+			}
+
+		}
+
+		klog.V(6).Infof("Prior update container resources containerCPUQuotaNew %.2f,containerCPUQuota %.2f,containerCPUPeriod %.2f,ContainerCPUUsages %.2f",
+			containerCPUQuotaNew, containerCPUQuota.Value, containerCPUPeriod.Value, v.Value)
+
+		if !utils.AlmostEqual(containerCPUQuotaNew*containerCPUPeriod.Value, containerCPUQuota.Value) {
```

Comment:
It is recommended to encapsulate `!utils.AlmostEqual(containerCPUQuotaNew*containerCPUPeriod.Value, containerCPUQuota.Value)` as a separate function